ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -697,21 +700,79 @@ public void findEntriesSinceOnRestartTest() throws IOException, StoreException {
     long bytesRead =
         log.getDifference(firstRecordFileSpan.getEndOffset(), new Offset(log.getFirstSegment().getName(), 0));
     // create a token that will be past the index end offset on startup after recovery.
-    startToken = new StoreFindToken(secondRecordFileSpan.getEndOffset(), oldSessionId);
+    startToken = new StoreFindToken(secondRecordFileSpan.getEndOffset(), oldSessionId, incarnationId, false);
     // token should get reset internally, no keys should be returned and the returned token should be correct (offset in
     // it will be the current log end offset = firstRecordFileSpan.getEndOffset()).
-    StoreFindToken expectedEndToken = new StoreFindToken(firstRecordFileSpan.getEndOffset(), sessionId);
+    StoreFindToken expectedEndToken =
+        new StoreFindToken(firstRecordFileSpan.getEndOffset(), sessionId, incarnationId, true);
     expectedEndToken.setBytesRead(bytesRead);
     doFindEntriesSinceTest(startToken, Long.MAX_VALUE, Collections.EMPTY_SET, expectedEndToken);
 
     // create a token that is not past the index end offset on startup after recovery. Should work as expected
-    startToken = new StoreFindToken(lastRecordOffset, oldSessionId);
-    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId);
+    startToken = new StoreFindToken(lastRecordOffset, oldSessionId, incarnationId, false);
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
     expectedEndToken.setBytesRead(bytesRead);
     doFindEntriesSinceTest(startToken, Long.MAX_VALUE, Collections.singleton(newId), expectedEndToken);
   }
 
   /**
+   * Tests behaviour of {@link PersistentIndex#findEntriesSince(FindToken, long)} on restart of index with a corner case
+   * scenario
+   * @throws IOException
+   * @throws StoreException
+   */
+  @Test
+  public void findEntriesSinceOnRestartCornerCaseTest() throws IOException, StoreException {
+    UUID oldSessionId = sessionId;
+    // ensure that index and log end offsets are in sync
+    addPutEntries(3, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    // test case description:
+    // after restart, the store(no new writes have gone into the store) will return a journal based token referring to
+    // logEndOffsetOnStartUp for a findEntriesSince(offset beyond logEndOffsetOnStartup with different session id) call
+    // On the subsequent findEntriesSince() call, the index should start returning entries
+    // starting from that offset and should not consider it as non-inclusive offset
+    recovery = new MessageStoreRecovery() {
+      @Override
+      public List<MessageInfo> recover(Read read, long startOffset, long endOffset, StoreKeyFactory factory)
+          throws IOException {
+        return Collections.emptyList();
+      }
+    };
+    reloadIndex(true);
+    Offset lastRecordOffset = index.journal.getLastOffset();
+    long bytesRead =
```

Comment:
can you put this line where you create the expected token? In this position, it is not clear why this variable is needed.